packages/itwin/tree-widget/src/components/trees/stateless/models-tree/internal/ModelsTreeQueryHandler.ts

```diff
@@ -0,0 +1,342 @@
+  rootElementIds: Id64String | Id64Set;
+}
+
+/**
+ * @internal
+ */
+export type ElementsQueryProps = ModelElementsQueryProps | CategoryElementsQueryProps | ElementsByParentQueryProps;
+
+export interface ElementInfo {
+  parentId?: Id64String;
+  elementId: Id64String;
+  modelId: Id64String;
+  categoryId: Id64String;
+}
+
+/**
+ * @internal
+ */
+export interface ModelsTreeQueryHandler {
+  /** Retrieves all models under a given subject */
+  querySubjectModels(subjectId: Id64String): Observable<Id64String>;
+  /** Retrieves all unique categories of elements under a given model */
+  queryModelCategories(modelId: Id64String): Observable<Id64String>;
+  /**
+   * Retrieves all elements that match given model, category and root (parent) element IDs,
+   * then recursively retrieves all children of those elements by the "ElementOwnsChildElements" relationship.
+   */
+  queryElements(props: ElementsQueryProps): Observable<Id64String>;
+  /** Analogous to `queryElements` but returns a count instead of values */
+  queryElementsCount(props: ElementsQueryProps): Observable<number>;
+  /** Returns information about given elements */
+  queryElementInfo(props: { elementIds: Id64String | Id64Set; useRootElementCategoryId?: boolean }): Observable<ElementInfo>;
+  /** Clears cached query results */
+  invalidateCache(): void;
+}
+
+/**
+ * @internal
+ */
+export function createModelsTreeQueryHandler(iModel: IModelConnection): ModelsTreeQueryHandler {
+  return new QueryHandlerImplementation(iModel);
+}
+
+class QueryHandlerImplementation implements ModelsTreeQueryHandler {
+  private readonly _groupedElementIdsCache = new Map<string, Observable<Id64String>>();
+  private readonly _elementsCountCache = new Map<string, Observable<number>>();
+  private _initialInfoObs?: Observable<InitialInfo>;
+
+  constructor(private readonly _iModel: IModelConnection) {}
+
+  public invalidateCache(): void {
+    this._groupedElementIdsCache.clear();
+    this._elementsCountCache.clear();
+    this._initialInfoObs = undefined;
+  }
+
+  public querySubjectModels(subjectId: string): Observable<string> {
+    return this.initialInfoObs.pipe(
+      map((state) => ({ ...state, modelIds: new Array<string>(), subjectId })),
+      expand((state) => {
+        const subjectModelIds = state.subjectModels.get(state.subjectId);
+        subjectModelIds && state.modelIds.push(...subjectModelIds);
+
+        const childSubjectIds = state.subjectsHierarchy.get(state.subjectId);
+        return childSubjectIds ? from(childSubjectIds).pipe(map((cs) => ({ ...state, subjectId: cs }))) : EMPTY;
+      }),
+      last(),
+      mergeMap(({ modelIds }) => modelIds),
+    );
+  }
+
+  private get initialInfoObs(): Observable<InitialInfo> {
+    return (this._initialInfoObs ??= forkJoin({
+      subjects: this.runSubjectsQuery().pipe(toArray()),
+      models: this.runModelsQuery().pipe(toArray()),
+      modelCategories: this.runModelCategoriesQuery().pipe(toArray()),
```

Comment:
I think subjects and models queries should be separate from the model categories query. We need the former as soon as the hierarchy is created, but the latter may be needed only later, when we expand deeper into the hierarchy.